subxt/src/client/lightclient/background.rs

```diff
@@ -0,0 +1,439 @@
+            mpsc::UnboundedSender<Box<RawValue>>,
+        ),
+    >,
+    /// Map the subscription ID to the frontend `Sender`.
+    subscriptions: HashMap<SubscriptionId, mpsc::UnboundedSender<Box<RawValue>>>,
+}
+
+impl BackgroundTask {
+    /// Constructs a new [`BackgroundTask`].
+    pub fn new(client: smoldot_light::Client<Arc<Platform>>, chain_id: ChainId) -> BackgroundTask {
+        BackgroundTask {
+            client,
+            chain_id,
+            request_id: 1,
+            requests: Default::default(),
+            id_to_subscription: Default::default(),
+            subscriptions: Default::default(),
+        }
+    }
+
+    /// Fetch and increment the request ID.
+    fn next_id(&mut self) -> usize {
+        let next = self.request_id;
+        self.request_id = self.request_id.wrapping_add(1);
+        next
+    }
+
+    /// Handle the registration messages received from the user.
+    async fn handle_requests(&mut self, message: FromSubxt) {
+        match message {
+            FromSubxt::Request {
+                method,
+                params,
+                sender,
+            } => {
+                let id = self.next_id();
+                let request = format!(
+                    r#"{{"jsonrpc":"2.0","id":"{}", "method":"{}","params":{}}}"#,
+                    id, method, params
+                );
+
+                self.requests.insert(id, sender);
+
+                let result = self.client.json_rpc_request(request, self.chain_id);
+                if let Err(err) = result {
+                    tracing::warn!(
+                        target: LOG_TARGET,
+                        "Cannot send RPC request to lightclient {:?}",
+                        err.to_string()
+                    );
+                    let sender = self
+                        .requests
+                        .remove(&id)
+                        .expect("Channel is inserted above; qed");
+
+                    // Send the error back to frontend.
+                    if sender
+                        .send(Err(LightClientError::Request(err.to_string())))
+                        .is_err()
+                    {
+                        tracing::warn!(
+                            target: LOG_TARGET,
+                            "Cannot send RPC request error to id={id}",
+                        );
+                    }
+                }
+            }
+            FromSubxt::Subscription {
+                method,
+                params,
+                sub_id,
+                sender,
+            } => {
+                // For subscriptions we need to make a plain RPC request to the subscription method.
+                // The server will return as a result the subscription ID.
+                let id = self.next_id();
+                let request = format!(
+                    r#"{{"jsonrpc":"2.0","id":"{}", "method":"{}","params":{}}}"#,
+                    id, method, params
+                );
```

Comment:
I think constructing the `request` variable and incrementing the id could be done before the match arms to avoid code duplication:
```
let id = self.next_id();
let (FromSubxt::Request { method, params, .. } | FromSubxt::Subscription { method, params, .. } )= &message;
let request = format!(
    r#"{{"jsonrpc":"2.0","id":"{}", "method":"{}","params":{}}}"#,
    id, method, params
);
```